File: Data Structures & Algorithms/validate-parentheses/submission-6.py
class Solution:
    def isValid(self, s: str) -> bool:
        par = {
            ')': '(',
            ']': '[',
            '}': '{'
        }

        stack = []

        for p in s:
            if p in par:
                if stack and stack[-1] == par[p]:
                    stack.pop()
                else:
                    return False

            else:
                stack.append(p)
        
        return len(stack) == 0

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-6.ts
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    largestRectangleArea(heights: number[]): number {
        let maxArea = 0
        const stack: [number, number][] = [] //[index, height]

        for (let i = 0; i < heights.length; i++) {
            const h = heights[i]
            let start = i

            while (stack.length && stack[stack.length - 1][1] > h) {
                const [index, height] = stack.pop()
                maxArea = Math.max(maxArea, height * (i - index))
                start = index
            }
            stack.push([start, h])
        }

        for (const [i, h] of stack) {
            maxArea = Math.max(maxArea, h * (heights.length - i))
        }

        return maxArea
    }
}
